[15.0][ADD] heldespk_mgmt: Helpdesk stages per teams

carolinafernandez-tecnativa opened the pull request:
Include a field in the stages to define the team. If it is empty, the stage will be available to all teams. If you have a team indicated, that stage will only be visible to tickets that belong to that team.

@Tecnativa
TT45710

@pedrobaeza @victoralmau

carolinafernandez-tecnativa committed:
[IMP] heldespk_mgmt: Helpdesk stages per teams

TT45710

carolinafernandez-tecnativa commented:
ping @pedrobaeza @victoralmau

pedrobaeza commented on helpdesk_mgmt/controllers/main.py:
```diff
@@ -84,11 +84,6 @@ def _prepare_submit_ticket_vals(self, **kw):
             "partner_id": request.env.user.partner_id.id,
             "partner_name": request.env.user.partner_id.name,
             "partner_email": request.env.user.partner_id.email,
-            # Need to set stage_id so that the _track_template() method is called
```
Why removing this?

carolinafernandez-tecnativa replied:
My mistake, it is not necessary to remove it.

pedrobaeza commented on helpdesk_mgmt/controllers/myaccount.py:
```diff
@@ -65,7 +65,13 @@ def portal_my_tickets(
-        for stage in request.env["helpdesk.ticket.stage"].search([]):
+        for stage in request.env["helpdesk.ticket.stage"].search(
+            [
+                "|",
+                ("team_ids", "=", HelpdeskTicket.team_id.id),
+                ("team_ids", "=", False),
```
You can do it in one shot:

```suggestion
                ("team_ids", "in", [False] + HelpdeskTicket.team_id.ids),
```

But anyway, that's incorrect. `HelpdeskTicket` is an empty recordset, so this is never going to have a team. You should have the team elsewhere.

pedrobaeza replied:
This suggestion is still not applied and I don't see a central method to get this. Something like `def _get_applicable_stages(self, team):` or similar.

carolinafernandez-tecnativa replied:
I defined _get_applicable_stages in portal and in helpdesk.ticket in order to get central method. Tell me if you see it okay now, thanks :)

pedrobaeza commented on helpdesk_mgmt/controllers/myaccount.py:
```diff
@@ -183,8 +189,15 @@ def portal_my_ticket(self, ticket_id, access_token=None, **kw):
-            [("close_from_portal", "=", True)]
+            [
+                "&",
+                "|",
+                ("team_ids", "=", ticket.team_id.id),
+                ("team_ids", "=", False),
```
```suggestion
                ("team_ids", "in", [False] + ticket.team_id.ids),
```

pedrobaeza commented on helpdesk_mgmt/controllers/myaccount.py:
```diff
@@ -183,8 +189,15 @@ def portal_my_ticket(self, ticket_id, access_token=None, **kw):
+                ("close_from_portal", "=", True),
+            ]
         )
+
```
Don't add empty lines inside a method.

pedrobaeza reviewed (changes requested):
You should try to have only one method for getting the proper stages, and call it everywhere instead of being doing changes in a lot of places. Remember also to put docstrings where applicable (like `_determine_stage`).

Finally, the statusbar should only contain the proper stages for the team.

carolinafernandez-tecnativa commented:
> You should try to have only one method for getting the proper stages, and call it everywhere instead of being doing changes in a lot of places. Remember also to put docstrings where applicable (like `_determine_stage`).
> 
> Finally, the statusbar should only contain the proper stages for the team.

Thanks for the review. could you please check it now?

carolinafernandez-tecnativa commented:
ping @pedrobaeza

carolinafernandez-tecnativa commented:
ping @pedrobaeza

pedrobaeza commented on helpdesk_mgmt/controllers/myaccount.py:
```diff
@@ -65,7 +65,8 @@ def portal_my_tickets(
             "all": {"label": _("All"), "domain": []},
         }
-        for stage in request.env["helpdesk.ticket.stage"].search([]):
+        stages = self._get_applicable_stages(HelpdeskTicket.team_id, False)
```
`HelpdeskTicket.team_id` is always going to be False, as there's no recordset. Why using this weird form?

pedrobaeza commented on helpdesk_mgmt/controllers/myaccount.py:
```diff
@@ -259,3 +258,13 @@ def _ticket_get_order(self, order, groupby):
+        domain = [
+            ("team_ids", "in", [False] + team.ids),
+        ]
+        if close_from_portal:
```
Instead of passing this argument (which by the way the implementation is not correct, as it shouldn't be with OR, just filtered out in the proper place.

pedrobaeza replied:
And use only one method in one model. Don't put it twice.

pedrobaeza commented on helpdesk_mgmt/controllers/myaccount.py:
```diff
@@ -259,3 +258,13 @@ def _ticket_get_order(self, order, groupby):
+        if close_from_portal:
+            domain = ["|"] + domain
+            domain.append(("close_from_portal", "=", True))
+
```
Don't leave empty line spaces inside a method.

pedrobaeza commented on helpdesk_mgmt/models/helpdesk_ticket.py:
```diff
@@ -10,13 +10,45 @@ class HelpdeskTicket(models.Model):
+        else:
+            return (
+                self.env["helpdesk.ticket.stage"]
+                .search([("team_ids", "=", False)], limit=1)
```
Why returning only one?

pedrobaeza commented on helpdesk_mgmt/models/helpdesk_ticket.py:
```diff
@@ -10,13 +10,45 @@ class HelpdeskTicket(models.Model):
+            ticket.stage_id = self._get_applicable_stages(ticket.team_id)
+
+    def _default_team_id(self):
+        team_id = (
```
This should call to the default team instead.

carolinafernandez-tecnativa replied:
what are you trying to say?

pedrobaeza replied:
Instead of repeating the code for the default team, call the method for getting the default team

pedrobaeza commented on helpdesk_mgmt/models/helpdesk_ticket.py:
```diff
@@ -10,13 +10,45 @@ class HelpdeskTicket(models.Model):
+        return self._get_applicable_stages(self.team_id)
+
+    @api.depends("team_id")
+    def _compute_stage_ids(self):
```
`_compute_stage_ids` (with `s`), but the assigned field is `stage_id`. Anyway, is this using the same technique as Odoo with stages in tasks?

carolinafernandez-tecnativa replied:
Yes, I renamed the method to _compute_stages_ids

pedrobaeza replied:
But the proper field name should be stage_ids

carolinafernandez-tecnativa commented:
WIP

pedrobaeza committed:
[IMP] helpdesk_mgmt: Simplify logic

- Remove default
- Centralize method for determining applicable stages
- Make compatible with multi-company

pedrobaeza committed:
[FIX] helpdesk_mgmt_project: Don't repeat tests

The inheritance should be over the base test class

pedrobaeza reviewed (approved):
Please check about the mail stuff, and it's good to go.

carolinafernandez-tecnativa commented:
> Please check about the mail stuff, and it's good to go.

I added this fix https://github.com/OCA/helpdesk/pull/522/files#diff-8e8d21502284f8964232135dd1c317a25eacbc0e9cae741627caf414b751498eR89 in order to receive mails when ticket is created from portal

pedrobaeza commented:
Let's go finally with this:

/ocabot merge major

pedrobaeza commented:
@carolinafernandez-tecnativa please fw-port it to v16/v17

carolinafernandez-tecnativa commented:
@pedrobaeza added on v16. https://github.com/OCA/helpdesk/pull/534

helpdesk_mgmt on V17 is not yet migrated.

pedrobaeza commented:
Write on the current opened PR to ask for its inclusion.

carolinafernandez-tecnativa commented:
> Write on the current opened PR to ask for its inclusion.

done!
